### DISCUSSION TITLE
Work started on client side LiteDb implementation

### DISCUSSION THREAD

**Discussion by adrianhall** [Azure/azure-mobile-apps#768] [General]
This implementation will take a few weeks.  However, if any of you wish to pull down the (currently non-functional) code, run the tests, play around, tell me where I'm going wrong, etc.  The code is in my personal repo here: https://github.com/adrianhall/azure-mobile-apps/tree/issues/356

If you wish to make comments on the functionality, then please do so as a comment to this thread.

Thanks all!

**Comment by dylan-mills** [comment]
Glad to hear this is getting worked on, are you still planning on doing the following mentioned in #624? 

![image](https://github.com/Azure/azure-mobile-apps/assets/6576706/5b3f06ef-8f48-457b-9dfc-b66a2c12ad8f)

I started playing around with creating a LiteDb client-side library earlier in the year and found dealing with the JObjects and BSON objects to be challenging. After looking at the Cosmos server library, which uses generics, I think it be would much easier to implement LiteDb without JObjects. That said, there may be a design reason for continuing to support JObject on the client side.

**Reply by adrianhall** [reply]
The short version for "design reason" is that the original code used JObject and without switching to a dictionary type structure (of which JObject is one), handling the query language is "hard".  It's also a complete re-write of the library, which is not a small thing.  It would be a complete breaking change.

That being said, it's something I want to pursue.  There are just too many other things on my plate right now.

**Reply by dylan-mills** [reply]
Thanks for the update, totally understand that kind of "design reason" 😂

I'll be happy to help do some testing when I get a chance!

**Comment by adrianhall** [comment]
More updates today - basic insertions, deletions, and querying is done.  I've also made a start on the query parser, which turns the QueryDescription into a BsonExpression.  Right now, I'm done with binary operators, constants(*), and string processing.

Still to do:
* ConvertNode
* UnaryNode
* Date/Time handlers
* Lots of testing
*

**Reply by adrianhall** [reply]
Update for today - I've mostly been working on completing the QueryFormatter and extending tests.  The QueryFormatter now handles all of the LiteDB syntax with the exception of the Unary "not" boolean operator.  I have to figure that one out.  I've extended the tests to show that all filters are workings.

Next up:

1. Figure out the Unary NOT filter syntax.
2. Add tests for selections, skip & top, and order by functionality.

And a little further out, add the LiteDb client-side to the integration tests, together with figuring out what is wrong when (inevitably) 80% of them will fail with the first run.  However, it looks like the hard work is done now.

**Reply by adrianhall** [reply]
... and just like that, I've bumped into an issue that is a blocker and cannot be fixed.

The short version is that we pass around a QueryDescription, which is a broken down semi-parsed IQueryable that allows us to convert into any form we want.  No problem there.  Unfortunately, we have to turn that into a LiteDB.Query, which is the base thing that the ILiteQueryable type generates.  In order to turn the QueryDescription into a LiteDB.Query, we need to parse the individual OData segments ($select, $orderBy, etc.) into BsonExpressions.

Internally, LiteDB does this.  But it's all internal.

So, I can either copy the entirety of LiteDB into my library (and maintain it going forward), or I have to give up on this idea.

I am not volunteering to maintain LiteDB going forward, so I am going to can this project.  It will remain on my GitHub repo (if someone wants to take it over), and I can discuss what still needs to be done, but copying code is not on the cards.
